composer/trainer/dist_strategy.py

```diff
@@ -0,0 +1,261 @@
+                                            ShardingStrategy)
+
+    sharding_map = {
+        'NO_SHARD': ShardingStrategy.NO_SHARD,
+        'SHARD_GRAD_OP': ShardingStrategy.SHARD_GRAD_OP,
+        'FULL_SHARD': ShardingStrategy.FULL_SHARD,
+    }
+    sharding_strategy = sharding_map[fsdp_config.get('sharding_strategy', 'FULL_SHARD').upper()]
+
+    cpu_offload = CPUOffload(offload_params=True) if fsdp_config.get('cpu_offload', False) else None
+    if cpu_offload is not None:
+        raise ValueError('FSDP CPU Offload not supported yet.')
+
+    def _get_torch_dtype(dtype: Union[Precision, str]):
+        dtype = dtype.value if isinstance(dtype, Precision) else dtype
+        if dtype in ['float32', 'torch.float32', 'fp32']:
+            return torch.float32
+        elif dtype in ['float16', 'torch.float16', 'half', 'fp16', 'amp', 'amp_fp16']:
+            return torch.float16
+        elif dtype in ['bfloat16', 'bfloat', 'torch.bfloat16', 'bf16', 'amp_bf16']:
+            return torch.bfloat16
+        else:
+            raise ValueError(f'Not sure how to convert dtype={dtype} to a torch dtype.')
+
+    if fsdp_config.get('mixed_precision', None):
+        param_dtype = _get_torch_dtype(fsdp_config['mixed_precision']['param_dtype'])
+        reduce_dtype = _get_torch_dtype(fsdp_config['mixed_precision']['reduce_dtype'])
+        buffer_dtype = _get_torch_dtype(fsdp_config['mixed_precision']['buffer_dtype'])
+    else:
+        param_dtype = torch.float32 if fsdp_config.get('fp32_master_weights', True) else _get_torch_dtype(precision)
+        reduce_dtype = _get_torch_dtype(precision)
+        buffer_dtype = param_dtype
+
+    mixed_precision = MixedPrecision(
+        param_dtype=param_dtype,
+        reduce_dtype=reduce_dtype,
+        buffer_dtype=buffer_dtype,
+    )
+    backward_prefetch = BackwardPrefetch.BACKWARD_POST
+
+    # We choose to not wrap the ComposerModel directly, but instead wrap any submodules like `ComposerModel.model`
+    # This makes it safer to call ComposerModel-specific functions like 'eval_forward' that
+    # may make calls to sharded submodules. If we only wrap the submodules, then any call that ComposerModel makes
+    # to a FSDP-wrapped submodule's `forward()` function will be safe and all-gather the necessary weights before `forward()`.
+    for name, obj in model.named_children():
+        if not isinstance(obj, (Metric, MetricCollection)):
+
+            # If `obj` contains meta tensors, try to use `obj.param_init_fn` to initialize them
+            def _param_init_fn(module: torch.nn.Module) -> None:
+                module.to_empty(device=f'cuda:{torch.cuda.current_device()}')
+                if hasattr(obj, 'param_init_fn') and isinstance(obj.param_init_fn, Callable):
+                    module.apply(obj.param_init_fn)
+                elif hasattr(module, 'reset_parameters') and isinstance(module.reset_parameters, Callable):
+                    module.reset_parameters()
```

Comment:
same minor nit on nested functions

Reply:
This one is gonna be a bit tricky because it's really defining a different wrap function per submodule `obj`, which is inside the for loop.. I think I'd like to leave as is and maybe we can refactor after?